UPGRADE_NOTES.md

```diff
@@ -1,3 +1,151 @@
+## Upgrading from 0.17 to 0.18
+
+- Upgrade to crystal 0.31.1
+- Upgrade Lucky CLI (homebrew)
+
+```
+brew update
+brew upgrade crystal-lang # Make sure you're up-to-date. Requires 0.31.1
+brew upgrade lucky
+```
+
+- Upgrade Lucky CLI (Linux)
+
+- Update `.crystal-version` to `0.31.1`
+
+> Remove the existing Lucky binary and follow the Linux
+> instructions in this section
+> https://luckyframework.org/guides/getting-started/installing#on-linux
+
+- Update versions in `shard.yml`
+  - Lucky should be `~> 0.18`
+- Run `shards update`
+
+### General updates
+
+- Rename: all `render` calls in actions to `html`.
+- Update: the `src/actions/errors/show.cr` file to the new format
+<details>
+  <summary>src/actions/errors/show.cr</summary>
+
+  ```crystal
+  class Errors::Show < Lucky::ErrorAction
+    DEFAULT_MESSAGE = "Something went wrong."
+    default_format :html
+    dont_report [Lucky::RouteNotFoundError]
+
+    def render(error : Lucky::RouteNotFoundError)
+      if html?
+        error_html "Sorry, we couldn't find that page.", status: 404
+      else
+        error_json "Not found", status: 404
+      end
+    end
+
+    # When the request is JSON and an InvalidOperationError is raised, show a
+    # helpful error with the param that is invalid, and what was wrong with it.
+    def render(error : Avram::InvalidOperationError)
+      if html?
+        error_html DEFAULT_MESSAGE, status: 500
+      else
+        error_json \
+          message: error.renderable_message,
+          details: error.renderable_details,
+          param: error.invalid_attribute_name,
+          status: 400
+      end
+    end
+
+    # Always keep this below other 'render' methods or it may override your
+    # custom 'render' methods.
+    def render(error : Lucky::RenderableError)
+      if html?
+        error_html DEFAULT_MESSAGE, status: error.renderable_status
+      else
+        error_json error.renderable_message, status: error.renderable_status
+      end
+    end
+
+    # If none of the 'render' methods return a response for the raised Exception,
+    # Lucky will use this method.
+    def default_render(error : Exception) : Lucky::Response
+      if html?
+        error_html DEFAULT_MESSAGE, status: 500
+      else
+        error_json DEFAULT_MESSAGE, status: 500
+      end
+    end
+
+    private def error_html(message : String, status : Int)
+      context.response.status_code = status
+      html Errors::ShowPage, message: message, status: status
+    end
+
+    private def error_json(message : String, status : Int, details = nil, param = nil)
+      json ErrorSerializer.new(message: message, details: details, param: param), status: status
+    end
+
+    private def report(error : Exception)
+      # Send to Rollbar, send an email, etc.
+    end
+  end
+  ```
+</details>
+- Rename: `title` to `message` in `src/pages/errors/show_page.cr`.
+- Add: `BaseSerializer` to `src/serializers/`.
```

Comment:
Also just realized  you need to require  this in src/app.cr

https://github.com/luckyframework/lucky_cli/blob/e3a98ed6d39663c24b82a0d8d71d1758cfdede84/src/web_app_skeleton/src/app.cr.ecr#L16